File: ahoy-server/src/main/java/za/co/lsd/ahoy/server/environments/EnvironmentService.java
/*
 * Copyright  2021 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package za.co.lsd.ahoy.server.environments;

import lombok.extern.slf4j.Slf4j;
import org.springframework.data.rest.core.annotation.RepositoryEventHandler;
import org.springframework.data.rest.webmvc.ResourceNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import za.co.lsd.ahoy.server.ReleaseService;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentRelease;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentReleaseId;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentReleaseRepository;

import java.util.List;

@Service
@RepositoryEventHandler
@Slf4j
public class EnvironmentService {
	private final EnvironmentRepository environmentRepository;
	private final EnvironmentReleaseRepository environmentReleaseRepository;
	private final ReleaseService releaseService;

	public EnvironmentService(EnvironmentRepository environmentRepository, EnvironmentReleaseRepository environmentReleaseRepository, ReleaseService releaseService) {
		this.environmentRepository = environmentRepository;
		this.environmentReleaseRepository = environmentReleaseRepository;
		this.releaseService = releaseService;
	}

	@Transactional
	public Environment create(Environment environment) {
		log.info("Creating environment: {}", environment);

		return environmentRepository.save(environment);
	}

	@Transactional
	public Environment destroy(Long environmentId) {
		Environment environment = environmentRepository.findById(environmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find environment: " + environmentId));

		return destroy(environment);
	}

	@Transactional
	public Environment destroy(Environment environment) {
		log.info("Destroying environment: {}", environment);

		undeployReleasesFrom(environment);

		environmentRepository.delete(environment);
		return environment;
	}

	@Transactional
	public Environment duplicate(Long sourceEnvironmentId, Long destEnvironmentId) {
		Environment sourceEnvironment = environmentRepository.findById(sourceEnvironmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find source environment: " + sourceEnvironmentId));

		Environment destEnvironment = environmentRepository.findById(destEnvironmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find destination environment: " + destEnvironmentId));

		log.info("Duplicating environment {} to environment {}", sourceEnvironment, destEnvironment);

		List<EnvironmentRelease> sourceEnvironmentReleases = sourceEnvironment.getEnvironmentReleases();
		for (EnvironmentRelease sourceEnvironmentRelease : sourceEnvironmentReleases) {

			EnvironmentRelease newEnvironmentRelease = new EnvironmentRelease();
			newEnvironmentRelease.setId(new EnvironmentReleaseId());
			newEnvironmentRelease.setRelease(sourceEnvironmentRelease.getRelease());
			newEnvironmentRelease.setEnvironment(destEnvironment);

			environmentReleaseRepository.save(newEnvironmentRelease);
		}
		return destEnvironment;
	}

	@Transactional
	public void updateOrderIndex(Long environmentId, Double orderIndex) {
		environmentRepository.updateOrderIndex(environmentId, orderIndex);
	}

	private void undeployReleasesFrom(Environment environment) {
		log.info("Checking if there are deployed releases in {}", environment.getName());
		environment.getEnvironmentReleases().forEach((environmentRelease) -> {
			if (environmentRelease.hasCurrentReleaseVersion()) {
				log.info("{} is currently deployed in {}, undeploying...", environmentRelease.getRelease().getName(), environment.getName());
				try {
					releaseService.undeploy(environmentRelease.getId()).get();
				} catch (Exception e) {
					throw new EnvironmentException("Failed to undeploy " + environmentRelease + " from " + environment, e);
				}
			}
		});
	}
}

File: ahoy-ui/src/app/environments/environment.service.ts
/*
 * Copyright  2021 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import {Injectable} from '@angular/core';
import {EMPTY, Observable} from 'rxjs';
import {catchError, map, tap} from 'rxjs/operators';
import {Notification} from '../notifications/notification';
import {NotificationsService} from '../notifications/notifications.service';
import {LoggerService} from '../util/logger.service';
import {RestClientService} from '../util/rest-client.service';
import {Environment} from './environment';

@Injectable({
  providedIn: 'root'
})
export class EnvironmentService {

  constructor(
    private log: LoggerService,
    private notificationsService: NotificationsService,
    private restClient: RestClientService) {
  }

  getAll(): Observable<Environment[]> {
    return this.restClient.get<any>('/data/environments?projection=environment&sort=orderIndex&sort=id').pipe(
      map(response => response._embedded.environments as Environment[]),
      tap((envs) => this.log.debug(`fetched ${envs.length} environments`))
    );
  }

  getAllForPromotion(releaseId: number): Observable<Environment[]> {
    const url = `/data/environments/search/forPromotion?releaseId=${releaseId}`;
    return this.restClient.get<any>(url).pipe(
      map(response => response._embedded.environments as Environment[]),
      tap((envs) => this.log.debug(`fetched ${envs.length} environments`))
    );
  }

  get(id: number): Observable<Environment> {
    const url = `/data/environments/${id}?projection=environment`;
    return this.restClient.get<Environment>(url).pipe(
      tap((env) => {
        this.log.debug('fetched environment', env);
      })
    );
  }

  create(environment: Environment): Observable<Environment> {
    this.log.debug('creating environment: ', environment);

    const url = `/data/environments/create`;

    return this.restClient.post<Environment>(url, environment, true).pipe(
      tap((createdEnvironment) => {
        this.log.debug('created environment', createdEnvironment);
        const text = `${createdEnvironment.name} ` + `was created in cluster ${createdEnvironment.cluster.name}`;
        this.notificationsService.notification(new Notification(text));
      }),
      catchError((error) => {
        const text = `Failed to create environment ${environment.name}`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  destroy(environment: Environment): Observable<Environment> {
    this.log.debug('destroying environment: ', environment);

    const id = environment.id;
    const url = `/data/environments/destroy/${id}`;

    return this.restClient.delete<Environment>(url, true).pipe(
      tap((destroyedEnvironment) => {
        this.log.debug('destroyed environment', environment);
        const text = `${destroyedEnvironment.name} ` + `was destroyed from cluster ${destroyedEnvironment.cluster.name}`;
        this.notificationsService.notification(new Notification(text));
      }),
      catchError((error) => {
        const text = `Failed to destroy environment ${environment.name}`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  duplicate(sourceEnvironment: Environment, destEnvironment: Environment): Observable<Environment> {
    this.log.debug('duplicating environment: ', sourceEnvironment);

    const url = `/data/environments/duplicate/${sourceEnvironment.id}/${destEnvironment.id}`;

    return this.restClient.post<Environment>(url, null, true).pipe(
      tap((duplicatedEnvironment) => {
        this.log.debug('duplicated environment', duplicatedEnvironment);
        const text = `${duplicatedEnvironment.name} ` + `was duplicated in cluster ${duplicatedEnvironment.cluster.name}`;
        this.notificationsService.notification(new Notification(text));
      }),
      catchError((error) => {
        const text = `Failed to duplicate environment ${destEnvironment.name} from environment: ${sourceEnvironment.name}`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  updateOrderIndex(environment: Environment): Observable<Environment> {
    this.log.debug('updating environment orderIndex: ', environment);

    const url = `/data/environments/${environment.id}/updateOrderIndex?orderIndex=${environment.orderIndex}`;

    return this.restClient.put<Environment>(url, null).pipe(
      tap(() => {
        this.log.debug('updated environment orderIndex', environment);
      }),
      catchError((error) => {
        const text = `Failed to update environment ${environment.name} order`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  link(id: number): string {
    return this.restClient.getLink('/data/environments', id);
  }
}
